mrc-6243 Compare to data

EmmaLRussell opened the pull request:
Adds `compareToData` method to DiscreteSystem and DiscreteSystemGenerator, with simple implementation in `discreteSIR` (not implemented in `discreteWalk`). 

Data can either be shared between all groups or separate data per group. If a single `TData` object, or a `TData[]` of length 1 is provided as a parameter, the data is assumed to be shared. If not, data is used per group and length of the data array must be equal to the number of groups. 

A `SystemDataComparison` class has been added to hold the comparison values, which can be accessed by group and particle index.

richfitz commented on src/DiscreteSystem.ts:
```diff
@@ -101,4 +102,35 @@ export class DiscreteSystem<TShared, TInternal> implements System {
+            const state = this._state.getParticle(iGroup, iParticle);
+            const shared = this._shared[iGroup];
+            const internal = this._internal[iGroup];
+            const comparisonValue = this._generator.compareData(
+                this._time,
+                particleStateToArray(state),
+                iData,
+                shared,
+                internal,
+                this._random
+            );
+            result.setValue(iGroup, iParticle, comparisonValue);
```
See comment in SystemDataComparison, but I don't see why this is not just writing into a vector/matrix directly in this function

EmmaLRussell commented on src/SystemDataComparison.ts:
```diff
@@ -0,0 +1,30 @@
+import { checkIntegerInRange } from "./utils.ts";
+
+export class SystemDataComparison {
```
Not needed, it just saves the consumer having to remember which way round the group and particle dimensions go, and is similar to accessors in e.g. SystemState. Can replace with ndarray or array of arrays.

EmmaLRussell commented on tests/DiscreteSystem.test.ts:
```diff
@@ -1,19 +1,37 @@
+    999999, // shared.N - shared.I0;
+    1, // shared.I0;
+    0,
+    0,
```
Added a placeholder ticket to the board: https://mrc-ide.myjetbrains.com/youtrack/issue/mrc-6257/Zero-every-functionality-for-incidence-calculations

EmmaLRussell commented:
I've moved `compareData` into a separate interface which extends the main generator interface, and subclassed the DiscreteSystem to make a Comparable variant of that too, which expects to use a comparable generator. So `discreteWalk` implements `DiscreteGenerator` while `discreteSIR` implements `ComparableDiscreteGenerator`. Because the comparable generator extends the base generator interface, you can give one to either a `DiscreteSystem` or `ComparableDiscreteSystem` to use for methods other than compare. 

Not sure if this was exactly what you were looking for!

Also did a bit of related housekeeping:
- shorted `DiscreteSystemGenerator` to `DiscreteGenerator` so that the new interface `ComparableDiscreteGenerator` isn't too ludicrously long.
- moved the main interfaces into their own folder to keep things tidy

Also - replace the comparison wrapper thing with ndarray for the compareData result type.

richfitz commented on src/ComparableDiscreteSystem.ts:
```diff
@@ -0,0 +1,55 @@
+
+export class ComparableDiscreteSystem<TShared, TInternal, TData>
+    extends DiscreteSystem<TShared, TInternal>
+    implements ComparableSystem<TData>
```
We can implement multiple interfaces, right? If so this feels exactly what we need

EmmaLRussell replied:
yep!
